firebase-components/src/main/java/com/google/firebase/components/Component.java

```diff
@@ -153,6 +168,12 @@ public boolean isValue() {
     return type == ComponentType.VALUE;
   }
 
+  /** Creates a copy of the component with {@link ComponentFactory} replaced. */
+  public Component<T> withFactory(ComponentFactory<T> factory) {
+    return new Component<>(
+        name, providedInterfaces, dependencies, instantiation, type, factory, publishedEvents);
+  }
+
```

Comment:
Is this for testing only? Looks like the purpose of `ComponentFactory` is only for product teams to implement. Once a product defined its factory in its registrars (and supply it via the component builder), it never needs to change?

Reply:
It's intended to be used outside of tests, for existing use example see: https://github.com/firebase/firebase-android-sdk/blob/da0f5937f964cfb2312f16b408285c44ae05840c/firebase-common/src/main/java/com/google/firebase/monitoring/ComponentMonitoring.java#L80-L84